Medable/ResearchStack
# PAT-1294: Review Step

### Objective
* Introduce ReviewStep as a Feature.

### Description
* The Backbone side of ReviewStep, consists of various changes, but it's tightly coupled with Axon, for both components work side-by-side.
* The build.gradle file has been updated to include the needed dependencies and add Kotlin features that were previously missing. 
* We are using the navigation component from Google to move around tasks. 
* ViewTaskActivity has been deprecated in favor of a new TaskActivity that uses Navigation. 
* A lot of model objects now implement `Cloneable` and equals/hashCode
* TaskResult has been modified to store the actual `Step` reference, not just the identifier. 
* `ReviewStepFragment` is the UI for reviewStep, it relies on a `ReviewStepLayout` to adhere to the existing Backbone architecture. 
* `ReviewStepLayout` is where the actual "view" lives (and the RecyclerView).
* Every Step is implemented as a simple Fragment that we navigate to using Navigation components. 
* Navigation is performed by constructing a `StepNavigationEvent`. This is the information needed to navigate. 
* The new `TaskActivity` heavily relies on a `TaskViewModel` that is shared by all steps. The activity's code is a copy from the old `ViewTaskActivity`, with customizations needed for ReviewStep.
* `TaskViewModel`'s core stuff is in the nextStep/prevStep() methods, where we must handle the logic to determine where to go (depending whether you're in Edit mode or not, and whether you're changing a branch, etc.).
* The Submit Bar has been updated to include a potential "cancel" button as well. (It's still not finished, buttons aren't correctly aligned)
* A bunch of new styles have been added, but nothing too fancy; abstractions around the margins/padding/etc. to avoid repeating the same numbers everywhere.
* A plethora of new XML files (all the steps' view holders) and fragments, etc. They are mostly called `rsb_STEP-TYPE_review_step_layout.xml` where `STEP-TYPE` can be any of `barcode_scanner` or `boolean`, etc. (Not happy with these names, I think we should have used something better, but we can rename them at some point, in the mean time, we don't care) :D
* 4-5 new strings (added to the translations Google Doc and with their locale prefixes (thanks @alaa-alshaikh )
* We have an abstract ViewHolder that is called `BaseReviewStepViewHolder` and it acts as the base for all steps ViewHolders.
* The `BaseReviewStepViewHolder` adds some extension functions for settings titles, questions, and answers (that can be given or "skipped")
* As you may have guessed there is a ViewHolder for each step type, they are mostly called `STEP-TYPE_StepViewHolder` where `STEP-TYPE` can be any step type (boolean, barcode_scanner, etc.)
* Each `STEP-TYPE_StepViewHolder` handles how it shows the title, question, and answer for this step (and any customizations needed, for example the ImageCapture has an image instead of answer)
 

### How to Test
In order to test ReviewStep, it's important to understand how it works. 

Review Step is a step type that you can add to any Survey task, preferably at the end (although this is not enforced by neither the web portal nor the actual code in the app), that allows you to see the list of steps you completed (or skipped) and go back, change the response, and continue. 

The step itself doesn't have a "value" or response, it's literally a RecyclerView that shows a cell/row for each step (that is not hidden or disabled, that is). 


* In order to see it working, we encourage you to use your existing tasks, add a step of the type `Review` at the end, and observe it. (It can be before the completion step, that's ok). 
* If you want an "make it for me" step, then you can play with this (no need to Log in, just Join Study and the task should appear): https://app.qa.medable.com/hybridstudy/axon/study/5dea5f25a4190b010019ecca

In essence, reach the review step, after answering a bunch of steps, and skipping some if you want you can do the following things: 

1. Tap the review step "edit" (pencil) button, go back and change the answer, save it and observe the value was updated.
2. Edit a value, but instead of "save" it, cancel (at the bottom, not the CANCEL at the top right), and observe you're taken back to the ReviewStep with the value **unmodified**.
3. If you have branches (complicated and important feature), you should change the value that triggered the branch change, and observe you are asked to complete all the things needed to fullfill the branching. 
4. Do the same as 3, but cancel before finishing, what happens? :)
5. Are the _copy_ (the actual strings) relevant? Do they make sense? are they confusing? Is an alert dialog hard to understand? Let us know. (I think some are, but I'm not saying which ones).
6. Submit the task at the end, did the values match what you had when you check on the web portal? did the task correctly upload? 

You get the idea. Don't be lazy and play with booleans only; don't be like me. ;)


### Known Bugs / Missing Features
* The Cancel button appears duplicated (top right and bottom left) when editing a step. It should only appear bottom left. [PAT-1448](https://jira.devops.medable.com/browse/PAT-1448) [**FIXED**]
* The Step's titles are wrong (it can say BooleanStepFragment) for example. [PAT-1444](https://jira.devops.medable.com/browse/PAT-1444)  [**FIXED**]
* Some copy is _confusing_. [PAT-1443](https://jira.devops.medable.com/browse/PAT-1443)
* There's some formatting issues in the code (spaces, brackets, missing or hard to understand comments) we're aware, we'll plan on improving this throughout the week.
* The Continuous scale value is reset to the value "-2147483648" whenever we take the step and one step after it, and then go back to the continuous scale step. [PAT-1446](https://jira.devops.medable.com/browse/PAT-1446) [**FIXED**]
* When changing a step's value that involves a branch change, there's no prompt [PAT-1443](https://jira.devops.medable.com/browse/PAT-1443)
